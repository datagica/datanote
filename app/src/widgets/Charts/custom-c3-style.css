.c3-chart-arc text {
  fill: rgba(255, 255, 255, 1);
  font-size: 15px;
  font-weight: 400;
  font-family: Roboto;
  letter-spacing: .4px;
  stroke-width: 0.2px;
  stroke-linecap: butt;
  stroke-linejoin: miter;
  text-shadow: 0px 0px 2px black;
}

.c3-legend-item {
  font-size: 11px;
  font-family: 'Roboto';
  font-weight: 300;
  text-transform: capitalize;
}


.c3-chart-arcs-title {
  dominant-baseline: middle;
  font-size: 11px;
  font-weight: 600;
  font-family: "Open Sans";
  letter-spacing: 0.2px;
  text-transform: uppercase;
  fill: rgba(0, 0, 0, 0.79);
}

.c3-charts-arc path {
  stroke: hsla(220,6%,90%,.91);
}
