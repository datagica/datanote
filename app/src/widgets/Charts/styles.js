const styles = {
  chart: { display: 'inline-block' },
  chartSize: {
    height: 200,
    width: 300
  },
  wideChartSize: {
    height: 200,
    width: 680, // 682px
  }
}
export default styles;
