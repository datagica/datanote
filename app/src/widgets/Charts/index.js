'use strict'

// fork from https://github.com/bcbcarl/react-c3js

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { findDOMNode } from 'react-dom'
import c3 from 'c3'
import 'c3/c3.css'
import './custom-c3-style.css'

const emptyFunc = () => {};

class Charts extends Component {
  static displayName = 'Charts';
  static propTypes = {
    data:        PropTypes.object.isRequired,
    title:       PropTypes.object,
    size:        PropTypes.object,
    padding:     PropTypes.object,
    color:       PropTypes.object,
    interaction: PropTypes.object,
    transition:  PropTypes.object,
    oninit:      PropTypes.func,
    onrendered:  PropTypes.func,
    onmouseover: PropTypes.func,
    onmouseout:  PropTypes.func,
    onresize:    PropTypes.func,
    onresized:   PropTypes.func,
    axis:        PropTypes.object,
    grid:        PropTypes.object,
    regions:     PropTypes.array,
    legend:      PropTypes.object,
    tooltip:     PropTypes.object,
    subchart:    PropTypes.object,
    zoom:        PropTypes.object,
    point:       PropTypes.object,
    line:        PropTypes.object,
    area:        PropTypes.object,
    bar:         PropTypes.object,
    pie:         PropTypes.object,
    donut:       PropTypes.object,
    gauge:       PropTypes.object,
    className:   PropTypes.string,
    style:       PropTypes.object
  };
  static defaultProps = {
    // no default for data
    title: {},
    size: {},
    padding: {},
    color: {},
    interaction: {},
    transition: {},
    oninit: emptyFunc,
    onrendered: emptyFunc,
    onmouseover: emptyFunc,
    onmouseout: emptyFunc,
    onresize: emptyFunc,
    onresized: emptyFunc,
    axis: {},
    grid: {},
    regions: [],
    legend: {},
    tooltip: {},
    subchart: {},
    zoom: {},
    point: {},
    line: {},
    area: {},
    bar: {},
    pie: {},
    donut: {},
    gauge: {}
  };

  componentWillReceiveProps(newProps) { this.updateChart(newProps)   }
  componentWillUnmount()              { this.destroyChart()          }
  componentDidMount()                 { this.updateChart(this.props) }

  generateChart(mountNode, config) {
    return c3.generate(Object.assign({ bindto: mountNode }, config))
  }

  destroyChart() {
    try {
      this.chart = this.chart.destroy()
    } catch (err) {
      throw new Error('Internal C3 error', err)
    }
  }

  updateChart(config) {
    if (this.chart) this.destroyChart();
    this.chart = this.generateChart(findDOMNode(this), Object.assign({}, config))
  }

  render() {
    return <div className={ this.props.className ? ` ${this.props.className}` : '' }
                    style={ this.props.style     ? this.props.style           : {} } />;
  }
}

export default Charts
