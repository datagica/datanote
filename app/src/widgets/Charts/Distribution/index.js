'use strict'

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ResponsiveLine } from '@nivo/line'
import styles from '~/widgets/Charts/styles'

// for distributions where X is continuous (eg. age distribution, hours etc..)
const Distribution = ({ label, data }) =>
  <div style={styles.wideChartSize}>
    <ResponsiveLine
        data={[
          {
            id: "",
            color: "hsl(253, 70%, 50%)",
            data: (console.log(data), data.map(({ x, value }) => ({
              x: x,
              y: value,
              // color: "hsl(40, 70%, 50%)",
            })))
          }
        ]}
        margin={{
          top: 50,
          right: 110,
          bottom: 50,
          left: 60
        }}
        minY="auto"
        stacked={true}
        curve="natural"
        axisBottom={{
          orient: "bottom",
          tickSize: 5,
          tickPadding: 5,
          tickRotation: 0,
          legend: label,
          legendOffset: 36,
          legendPosition: "center"
        }}
        axisLeft={{
          orient: "left",
          tickSize: 5,
          tickPadding: 5,
          tickRotation: 0,
          legend: "count",
          legendOffset: -40,
          legendPosition: "center"
        }}
        dotSize={9}
        dotColor="inherit:darker(0.3)"
        dotBorderWidth={2}
        dotBorderColor="#ffffff"
        enableDotLabel={false}
        dotLabel="y"
        dotLabelYOffset={-12}
        enableArea={true}
        areaOpacity={0.1}
        animate={true}
        motionStiffness={90}
        motionDamping={15}
        legends={[
          {
            anchor: "bottom-right",
            direction: "column",
            translateX: 100,
            itemWidth: 80,
            itemHeight: 20,
            symbolSize: 12,
            symbolShape: "circle"
          }
        ]}
    />
  </div>

export default Distribution
