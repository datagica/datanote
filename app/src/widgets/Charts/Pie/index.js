import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ResponsivePie } from '@nivo/pie'
import styles from '~/widgets/Charts/styles'
import capitalize from '~/utils/formatters/capitalize'


const Pie = ({ label, data }) =>
<div style={styles.chartSize}>
  <ResponsivePie
    data={Object.keys(data).map((key, i) => ({
      id: key,
      label: capitalize(key),
      value: data[key],
      /*
      problem of this approach, colors are not opposed..
      color: `hsl(${Math.round(
        i * (255 / Object.keys(data).length)
      )}, 70%, 50%)`
      */

    }))}

    margin={{
        "top": 15,
        "right": 40,
        "bottom": 30,
        "left": 40
    }}
    innerRadius={0.5}
    padAngle={0.7}
    cornerRadius={3}
    colors="d320c"
    colorBy="id"
    borderColor="inherit:darker(0.6)"
    radialLabelsSkipAngle={10}
    radialLabelsTextXOffset={6}
    radialLabelsTextColor="#333333"
    radialLabelsLinkOffset={0}
    radialLabelsLinkDiagonalLength={16}
    radialLabelsLinkHorizontalLength={24}
    radialLabelsLinkStrokeWidth={1}
    radialLabelsLinkColor="inherit"
    slicesLabelsSkipAngle={10}
    slicesLabelsTextColor="#333333"
    animate={true}
    motionStiffness={90}
    motionDamping={15}
    legends={[
      {
        anchor: "bottom",
        direction: "row",
        translateY: 56,
        itemWidth: 80,
        itemHeight: 12,
        symbolSize: 12,
        symbolShape: "circle"
      }
    ]}
  />
  </div>

export default Pie
