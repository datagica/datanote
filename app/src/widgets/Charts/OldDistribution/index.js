'use strict'

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Charts from '~/widgets/Charts'
import styles from '~/widgets/Charts/styles'

// for distributions where X is continuous (eg. age distribution, hours etc..)
const Distribution = ({ label, data }) => (
  <Charts
    key={label.toLowerCase()}
    style={ styles.chart }
    size={ styles.chartSize }
    axis={{
      x: {
        // 'category' is for discrete data
        // type: 'category',
      },

      // For continuous values (distributions) we leave the default, as it is
      // a rather classic numerical chart.
      // rotated: false
    }}
    bar={{
      width: {
        ratio: 0.8
      }
    }}
    data={(console.log(data), {
      type: 'bar',
      json: data, // makeContinuous(data),
      keys: {
        x: 'x',
        value: [ 'value' ]
      },
      onclick: (d, element) => {
        console.log("Clicked on bar: ", { d: d, element: element })
      }
    })}
  />
)

export default Distribution
