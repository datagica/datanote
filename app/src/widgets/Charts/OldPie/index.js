'use strict'

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Charts from '~/widgets/Charts'
import styles from '~/widgets/Charts/styles'
import capitalize from '~/utils/formatters/capitalize'

const Pie = ({ label, data }) => (
  <Charts
    key={label.toLowerCase()}
    style={ styles.chart }
    size={ styles.chartSize }
    donut={{
      title: capitalize(label),
      width: 42
    }}
    data={{
      type: 'donut',
      json: [ data ],
      keys: {
        value: Object.keys(data)
      },
      onclick: (d, element) => {
        console.log("Clicked on donut: ", { d: d, element: element })
      }
    }}
  />
)

export default Pie
