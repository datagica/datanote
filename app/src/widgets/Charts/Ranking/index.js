'use strict'

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Charts from '~/widgets/Charts'
import styles from '~/widgets/Charts/styles'

const Ranking = ({ label, data }) => (
  <Charts
    key={label.toLowerCase()}
    style={ styles.chart }
    size={ styles.chartSize }
    axis={{
      x: {
        type: 'category',
      },
      // For discrete values (enumerations) we rotate to have horizontal bars and
      // sorted list, since there can be many values.
      rotated: true
    }}
    bar={{
      width: {
        ratio: 0.8
      }
    }}
    data={{
      type: 'bar',
      json: data
        .sort((a, b) => b.value - a.value)
        .slice(0, 15), // no scroll yet so we have to crop the data
      keys: {
        x: 'x',
        value: [ 'value' ]
      },
      onclick: (d, element) => {
        console.log("Clicked on bar: ", { d: d, element: element })
      }
    }}
  />
)

export default Ranking
