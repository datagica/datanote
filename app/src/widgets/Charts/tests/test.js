
// Just some mocks for D3 and C3
jest.mock('react-dom');
window.MutationObserver = window.MutationObserver || function() {
  return { observe : function() {} };
};

import React    from 'react';
import renderer from 'react-test-renderer';

import Charts  from '../index';

test('Default Charts', () => {
  const component = renderer.create(
    <Charts
      style={{}}
      size={{
        width: 200,
        height: 200
      }}
      donut={{ title: "Test" }}
      data={{
        type: 'donut',
        json: [{
          x: 25,
          y: 25,
          z: 50
        }],
        keys: {
          value: [ 'x', 'y', 'z' ]
        },
        onclick: (d, element) => {}
      }}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // note: since nearly everything is mocked, we don't get much DOM out of the
  // component anyway.. still, it tests props parsing and stuff
});
