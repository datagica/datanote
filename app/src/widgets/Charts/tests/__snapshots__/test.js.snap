// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Charts 1`] = `
<div
  className=""
  style={Object {}}
/>
`;
